classes/migration/upgrade/v3_4_0/I7167_RemoveDuplicatedUserSettingsAndDeprecatedFields.inc.php

```diff
@@ -0,0 +1,114 @@
+                LEFT JOIN user_settings best
+                    ON best.setting_name = duplicated.setting_name
+                    AND best.user_id = duplicated.user_id
+                    AND best.locale = duplicated.locale
+                    AND (
+                        COALESCE(best.assoc_id, 999999) < COALESCE(duplicated.assoc_id, 999999)
+                        OR (
+                            COALESCE(best.assoc_id, 999999) = COALESCE(duplicated.assoc_id, 999999)
+                            AND COALESCE(best.assoc_type, 999999) < COALESCE(duplicated.assoc_type, 999999)
+                        )
+                    )
+                -- Ensures a better record was found (if not found, it means the current duplicated record is the best and shouldn't be removed)
+                WHERE best.user_id IS NOT NULL"
+            );
+        } else {
+            DB::unprepared(
+                "DELETE FROM user_settings s
+                USING user_settings duplicated
+                -- Attempts to find a better fitting record among the duplicates (preference is given to the smaller assoc_id/assoc_type values)
+                LEFT JOIN user_settings best
+                    ON best.setting_name = duplicated.setting_name
+                    AND best.user_id = duplicated.user_id
+                    AND best.locale = duplicated.locale
+                    AND (
+                        COALESCE(best.assoc_id, 999999) < COALESCE(duplicated.assoc_id, 999999)
+                        OR (
+                            COALESCE(best.assoc_id, 999999) = COALESCE(duplicated.assoc_id, 999999)
+                            AND COALESCE(best.assoc_type, 999999) < COALESCE(duplicated.assoc_type, 999999)
+                        )
+                    )
+                -- Locates all duplicated settings (same key fields, except the assoc_type/assoc_id)
+                WHERE s.setting_name = duplicated.setting_name
+                    AND s.user_id = duplicated.user_id
+                    AND s.locale = duplicated.locale
+                    AND (
+                        COALESCE(s.assoc_type, -999999) <> COALESCE(duplicated.assoc_type, -999999)
+                        OR COALESCE(s.assoc_id, -999999) <> COALESCE(duplicated.assoc_id, -999999)
+                    )
+                    -- Ensures a better record was found (if not found, it means the current duplicated record is the best and shouldn't be removed)
+                    AND best.user_id IS NOT NULL"
+            );
+        }
+
+        // Here we should be free of duplicates, so it's safe to remove the columns without creating duplicated entries.
+        Schema::table(
+            'user_settings',
+            function (Blueprint $table): void {
+                // Drop deprecated fields (this will implicitly drop the unique key "user_settings_pkey")
```

Comment:
From https://dev.mysql.com/doc/refman/8.0/en/alter-table.html:
> If columns are dropped from a table, the columns are also removed from any index of which they are a part. If all columns that make up an index are dropped, the index is dropped as well. If you use CHANGE or MODIFY to shorten a column for which an index exists on the column, and the resulting column length is less than the index length, MySQL shortens the index automatically. 

...so I think this will drop those columns from the `user_settings_pkey` index, but not delete the whole index.

As for PostgreSQL (https://www.postgresql.org/docs/7.4/sql-altertable.html):
> Indexes and table constraints involving the column will be automatically dropped as well.

This is awfully brief but seems to suggest that the index will be fully dropped.

Reply:
That's true! I was deceived by my MySQL editor (it's adding an extra clause to remove the index) 😁